contracts/MintNotes.sol

```diff
@@ -0,0 +1,79 @@
+pragma solidity ^0.4.25;
+
+import {Verifier as MintNoteVerifier} from "./verifiers/MintNoteVerifier.sol";
+import "./ZkDaiBase.sol";
+
+contract MintNotes is MintNoteVerifier, ZkDaiBase {
+  uint8 internal constant NUM_PUBLIC_INPUTS = 4;
+
+  /**
+  * @dev Hashes the submitted proof and adds it to the submissions mapping that tracks
+  *      submission time, type, public inputs of the zkSnark and the submitter
+  */
+  function submit(
+      uint[2] a,
+      uint[2] a_p,
+      uint[2][2] b,
+      uint[2] b_p,
+      uint[2] c,
+      uint[2] c_p,
+      uint[2] h,
+      uint[2] k,
+      uint[4] input)
+    internal {
+      bytes32 proofHash = getProofHash(a, a_p, b, b_p, c, c_p, h, k);
+      uint256[] memory publicInput = new uint256[](4);
+      for(uint8 i = 0; i < NUM_PUBLIC_INPUTS; i++) {
+        publicInput[i] = input[i];
+      }
+      submissions[proofHash] = Submission(msg.sender, SubmissionType.Mint, now, publicInput);
+      emit Submitted(msg.sender, proofHash);
```

Comment:
The entire proof should be emitted in the event so it is easy to fetch the proof data (otherwise if submitted through a smart contract it would only be visible by checking transaction traces).

Reply:
Yes, but wouldn't this approach save us loads in gas? Is it considered a bad design practice to  be checking transaction traces?  - i see this tradeoff as lower costs for added complexity on the client side (but since it's off-chain, it's doable). Thoughts?

Reply:
Yeah. It is more expensive to log everything but I would say it is worth it. Requiring traces is not a great design pattern IMO.

Reply:
apparently truffle has issues parsing emitted arrays.
```
Uncaught Error: Error: [number-to-bn] while converting number ["1664107541672151440648312920369544688533654672829066134244365264292154508432","20992914610679640420663765902577714960355608522971040254942020341103009578163"] to BN.js instance, error: invalid number value. Value must be an integer, hex string, BN or BigNumber instance. Note, decimals are not supported. Given value: "1664107541672151440648312920369544688533654672829066134244365264292154508432,20992914610679640420663765902577714960355608522971040254942020341103009578163"
```